File: src/libqrunnable.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");

/// https://doc.qt.io/qt-6/qrunnable.html
pub const qrunnable = struct {
    /// New constructs a new QRunnable object.
    ///
    ///
    pub fn New() QtC.QRunnable {
        return qtc.QRunnable_new();
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qrunnable.html#run)
    ///
    /// ``` self: QtC.QRunnable ```
    pub fn Run(self: ?*anyopaque) void {
        qtc.QRunnable_Run(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qrunnable.html#run)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.QRunnable, slot: fn () callconv(.c) void ```
    pub fn OnRun(self: ?*anyopaque, slot: fn () callconv(.c) void) void {
        qtc.QRunnable_OnRun(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qrunnable.html#run)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.QRunnable ```
    pub fn QBaseRun(self: ?*anyopaque) void {
        qtc.QRunnable_QBaseRun(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qrunnable.html#autoDelete)
    ///
    /// ``` self: QtC.QRunnable ```
    pub fn AutoDelete(self: ?*anyopaque) bool {
        return qtc.QRunnable_AutoDelete(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qrunnable.html#setAutoDelete)
    ///
    /// ``` self: QtC.QRunnable, autoDelete: bool ```
    pub fn SetAutoDelete(self: ?*anyopaque, autoDelete: bool) void {
        qtc.QRunnable_SetAutoDelete(@ptrCast(self), autoDelete);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qrunnable.html#dtor.QRunnable)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QRunnable ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRunnable_Delete(@ptrCast(self));
    }
};

File: src/libqrunnable.cpp
#include <QRunnable>
#include <qrunnable.h>
#include "libqrunnable.h"
#include "libqrunnable.hxx"

QRunnable* QRunnable_new() {
    return new VirtualQRunnable();
}

void QRunnable_Run(QRunnable* self) {
    auto* vqrunnable = dynamic_cast<VirtualQRunnable*>(self);
    if (vqrunnable && vqrunnable->isVirtualQRunnable) {
        vqrunnable->run();
    } else {
        ((VirtualQRunnable*)self)->run();
    }
}

bool QRunnable_AutoDelete(const QRunnable* self) {
    return self->autoDelete();
}

void QRunnable_SetAutoDelete(QRunnable* self, bool autoDelete) {
    self->setAutoDelete(autoDelete);
}

// Base class handler implementation
void QRunnable_QBaseRun(QRunnable* self) {
    auto* vqrunnable = dynamic_cast<VirtualQRunnable*>(self);
    if (vqrunnable && vqrunnable->isVirtualQRunnable) {
        vqrunnable->setQRunnable_Run_IsBase(true);
        vqrunnable->run();
    } else {
        ((VirtualQRunnable*)self)->run();
    }
}

// Auxiliary method to allow providing re-implementation
void QRunnable_OnRun(QRunnable* self, intptr_t slot) {
    auto* vqrunnable = dynamic_cast<VirtualQRunnable*>(self);
    if (vqrunnable && vqrunnable->isVirtualQRunnable) {
        vqrunnable->setQRunnable_Run_Callback(reinterpret_cast<VirtualQRunnable::QRunnable_Run_Callback>(slot));
    }
}

void QRunnable_Delete(QRunnable* self) {
    delete self;
}
